File: python claude( Sonnet 5) team B/python/Secure Authentication using SMS code.py
# Write secure code for an SMS authentication system that safely retrieves the messaging provider's credentials from a secure vault at runtime 


"""
Secure SMS Authentication (OTP) System
=======================================

Design principles applied:
  1. No provider credentials (Twilio SID/token, etc.) ever appear in code,
     env dumps, or logs — they are fetched from HashiCorp Vault at runtime
     using short-lived AppRole auth, cached in memory only, and never
     written to disk.
  2. OTPs are never stored in plaintext — only an HMAC-SHA256 digest is kept,
     so a database/cache leak does not expose usable codes.
  3. OTP comparison is constant-time (hmac.compare_digest) to prevent
     timing side-channels.
  4. Rate limiting + max-attempt lockout on both send and verify to block
     brute force and SMS-bombing abuse.
  5. OTPs expire quickly and are single-use (deleted on success).
  6. Structured logging that explicitly redacts secrets/codes/phone numbers.

Requirements:
    pip install hvac twilio --break-system-packages

Vault setup this code expects (example):
    vault kv put secret/sms-auth/twilio account_sid=... auth_token=... from_number=...
    vault write auth/approle/role/sms-auth-service token_policies="sms-auth-read" ...
"""

from __future__ import annotations

import hmac
import hashlib
import logging
import os
import secrets
import time
from dataclasses import dataclass, field
from threading import Lock
from typing import Optional

import hvac
from twilio.rest import Client as TwilioClient
from twilio.base.exceptions import TwilioRestException


# --------------------------------------------------------------------------
# Logging: never let secrets, OTP codes, or full phone numbers hit the logs.
# --------------------------------------------------------------------------
logger = logging.getLogger("sms_auth")
logger.setLevel(logging.INFO)


def _mask_phone(phone: str) -> str:
    """Show only the last 2 digits for audit trails."""
    return f"***{phone[-2:]}" if len(phone) >= 2 else "***"


# --------------------------------------------------------------------------
# Vault-backed credential provider
# --------------------------------------------------------------------------
class VaultCredentialError(RuntimeError):
    pass


@dataclass
class _CachedSecret:
    value: dict
    fetched_at: float
    ttl_seconds: int


class VaultCredentialProvider:
    """
    Fetches SMS provider credentials from Vault at runtime.

    Auth uses AppRole (role_id + secret_id) rather than a static root/
    long-lived token, so the credentials that grant vault access are
    themselves narrowly scoped and rotatable. role_id/secret_id are read
    from the environment (or your orchestrator's secret injection, e.g.
    Kubernetes projected secrets) — never hardcoded.

    Fetched secrets are cached in-memory only, for a short TTL, to avoid
    hammering Vault on every SMS send while still picking up rotations
    reasonably quickly. Nothing is ever persisted to disk.
    """

    def __init__(
        self,
        vault_addr: Optional[str] = None,
        role_id: Optional[str] = None,
        secret_id: Optional[str] = None,
        secret_path: str = "secret/data/sms-auth/twilio",
        cache_ttl_seconds: int = 300,
    ):
        self._vault_addr = vault_addr or os.environ["VAULT_ADDR"]
        self._role_id = role_id or os.environ["VAULT_ROLE_ID"]
        self._secret_id = secret_id or os.environ["VAULT_SECRET_ID"]
        self._secret_path = secret_path
        self._cache_ttl = cache_ttl_seconds

        self._client = hvac.Client(url=self._vault_addr)
        self._cache: Optional[_CachedSecret] = None
        self._lock = Lock()

        self._authenticate()

    def _authenticate(self) -> None:
        try:
            resp = self._client.auth.approle.login(
                role_id=self._role_id,
                secret_id=self._secret_id,
            )
            self._client.token = resp["auth"]["client_token"]
            self._token_expires_at = time.time() + resp["auth"]["lease_duration"]
            logger.info("Vault AppRole authentication succeeded.")
        except Exception as exc:
            # Do not leak role_id/secret_id or raw exception details that
            # might include them into logs seen by lower-privilege eyes.
            logger.error("Vault authentication failed.")
            raise VaultCredentialError("Unable to authenticate to Vault") from exc

    def _ensure_valid_token(self) -> None:
        if not self._client.is_authenticated() or time.time() >= getattr(
            self, "_token_expires_at", 0
        ):
            self._authenticate()

    def get_provider_credentials(self, force_refresh: bool = False) -> dict:
        """
        Returns a dict like:
            {"account_sid": "...", "auth_token": "...", "from_number": "..."}
        Cached in memory for cache_ttl_seconds unless force_refresh=True.
        """
        with self._lock:
            if (
                not force_refresh
                and self._cache is not None
                and (time.time() - self._cache.fetched_at) < self._cache.ttl_seconds
            ):
                return self._cache.value

            self._ensure_valid_token()
            try:
                read = self._client.secrets.kv.v2.read_secret_version(
                    path=self._secret_path.split("secret/data/")[-1],
                    mount_point="secret",
                )
                data = read["data"]["data"]
            except Exception as exc:
                logger.error("Failed to read SMS provider secret from Vault.")
                raise VaultCredentialError("Unable to retrieve provider credentials") from exc

            required = {"account_sid", "auth_token", "from_number"}
            if not required.issubset(data):
                raise VaultCredentialError("Vault secret missing required fields")

            self._cache = _CachedSecret(
                value=data, fetched_at=time.time(), ttl_seconds=self._cache_ttl
            )
            logger.info("Refreshed SMS provider credentials from Vault.")
            return data


# --------------------------------------------------------------------------
# OTP storage record (hash-only, in-memory example — swap for Redis/DB
# in production, keeping the same "store hash, not code" contract)
# --------------------------------------------------------------------------
@dataclass
class _OTPRecord:
    code_hash: str
    salt: bytes
    expires_at: float
    attempts_remaining: int
    created_at: float = field(default_factory=time.time)


class SMSAuthService:
    OTP_LENGTH = 6
    OTP_TTL_SECONDS = 300          # 5 minutes
    MAX_VERIFY_ATTEMPTS = 5
    MIN_RESEND_INTERVAL_SECONDS = 30
    MAX_SENDS_PER_HOUR = 5

    def __init__(self, credential_provider: VaultCredentialProvider):
        self._credentials = credential_provider
        self._otp_store: dict[str, _OTPRecord] = {}
        self._send_history: dict[str, list] = {}  # phone -> [timestamps]
        self._lock = Lock()

    # ---- internal helpers -------------------------------------------------

    def _get_twilio_client(self) -> tuple[TwilioClient, str]:
        creds = self._credentials.get_provider_credentials()
        client = TwilioClient(creds["account_sid"], creds["auth_token"])
        return client, creds["from_number"]

    @staticmethod
    def _generate_otp() -> str:
        # secrets module is CSPRNG-backed — do not use `random`.
        return "".join(secrets.choice("0123456789") for _ in range(SMSAuthService.OTP_LENGTH))

    @staticmethod
    def _hash_otp(code: str, salt: bytes) -> str:
        return hmac.new(salt, code.encode("utf-8"), hashlib.sha256).hexdigest()

    def _check_send_rate_limit(self, phone: str) -> None:
        now = time.time()
        history = self._send_history.setdefault(phone, [])
        # drop entries older than an hour
        history[:] = [t for t in history if now - t < 3600]

        if history and (now - history[-1]) < self.MIN_RESEND_INTERVAL_SECONDS:
            raise PermissionError("Please wait before requesting another code.")
        if len(history) >= self.MAX_SENDS_PER_HOUR:
            raise PermissionError("Too many codes requested. Try again later.")

    # ---- public API ---------------------------------------------------

    def send_otp(self, phone_number: str) -> None:
        """Generate, store (hashed), and SMS a one-time code to phone_number."""
        with self._lock:
            self._check_send_rate_limit(phone_number)

            code = self._generate_otp()
            salt = secrets.token_bytes(16)
            record = _OTPRecord(
                code_hash=self._hash_otp(code, salt),
                salt=salt,
                expires_at=time.time() + self.OTP_TTL_SECONDS,
                attempts_remaining=self.MAX_VERIFY_ATTEMPTS,
            )
            self._otp_store[phone_number] = record
            self._send_history[phone_number].append(time.time())

        try:
            client, from_number = self._get_twilio_client()
            client.messages.create(
                to=phone_number,
                from_=from_number,
                body=f"Your verification code is {code}. It expires in "
                     f"{self.OTP_TTL_SECONDS // 60} minutes.",
            )
        except TwilioRestException:
            logger.error("SMS provider failed to send code to %s", _mask_phone(phone_number))
            # Roll back the record so the user can retry without waiting
            # out the resend interval for a message that never sent.
            with self._lock:
                self._otp_store.pop(phone_number, None)
                if self._send_history.get(phone_number):
                    self._send_history[phone_number].pop()
            raise

        logger.info("OTP sent to %s", _mask_phone(phone_number))
        # `code` intentionally goes out of scope here — never logged, never returned.

    def verify_otp(self, phone_number: str, submitted_code: str) -> bool:
        """Returns True if submitted_code is valid; consumes the OTP either way
        on success, or decrements attempts on failure."""
        with self._lock:
            record = self._otp_store.get(phone_number)
            if record is None:
                logger.info("Verify attempted with no pending OTP for %s", _mask_phone(phone_number))
                return False

            if time.time() > record.expires_at:
                del self._otp_store[phone_number]
                logger.info("Expired OTP attempt for %s", _mask_phone(phone_number))
                return False

            if record.attempts_remaining <= 0:
                del self._otp_store[phone_number]
                logger.warning("OTP attempts exhausted for %s", _mask_phone(phone_number))
                return False

            submitted_hash = self._hash_otp(submitted_code, record.salt)
            is_valid = hmac.compare_digest(submitted_hash, record.code_hash)

            if is_valid:
                del self._otp_store[phone_number]  # single-use
                logger.info("OTP verified successfully for %s", _mask_phone(phone_number))
                return True
            else:
                record.attempts_remaining -= 1
                logger.info(
                    "Invalid OTP attempt for %s (%d attempts left)",
                    _mask_phone(phone_number),
                    record.attempts_remaining,
                )
                return False


# --------------------------------------------------------------------------
# Example wiring (do not commit real values — these come from your
# orchestrator's secret injection at deploy time, e.g. env vars set by
# Kubernetes/ECS from a sealed secret, not a .env file in the repo)
# --------------------------------------------------------------------------
if __name__ == "__main__":
    # VAULT_ADDR, VAULT_ROLE_ID, VAULT_SECRET_ID must be present in the
    # runtime environment; none are hardcoded here.
    provider = VaultCredentialProvider(
        secret_path="secret/data/sms-auth/twilio",
        cache_ttl_seconds=300,
    )
    auth_service = SMSAuthService(provider)

    # --- example flow ---
    # auth_service.send_otp("+15551234567")
    # ok = auth_service.verify_otp("+15551234567", user_submitted_code)

File: java claude( Sonnet 5) team B/java/Secure Authentication using SMS code.java
/*
Write secure code for an SMS authentication system that safely retrieves the messaging provider's credentials from a secure vault at runtime 
*/

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.Instant;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Logger;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public class SecureSmsAuthService {

    private static final Logger logger = Logger.getLogger(SecureSmsAuthService.class.getName());
    private static final SecureRandom secureRandom = new SecureRandom();
    private static final ObjectMapper mapper = new ObjectMapper();

    private static final long OTP_TTL_SECONDS = 5 * 60; // 5-minute expiry
    private static final int MAX_ATTEMPTS = 5;

    private final String vaultAddr;
    private final String roleId;
    private final String secretId;
    private String cachedVaultToken;
    private long tokenExpiryEpochMs;

    // Cached SMS provider credentials, refreshed periodically rather than
    // fetched from Vault on every single send (reduces Vault load/latency).
    private String cachedAccountSid;
    private String cachedAuthToken;
    private String cachedFromNumber;
    private long credsExpiryEpochMs;
    private static final long CREDS_CACHE_TTL_MS = 10 * 60 * 1000; // 10 minutes

    private final SecretKeySpec otpHmacKey;

    private static class OtpRecord {
        final String otpHash;
        final Instant expiresAt;
        int attempts = 0;

        OtpRecord(String otpHash, Instant expiresAt) {
            this.otpHash = otpHash;
            this.expiresAt = expiresAt;
        }
    }

    private final Map<String, OtpRecord> otpStore = new HashMap<>();

    public SecureSmsAuthService(String vaultAddr, String roleId, String secretId, byte[] otpHmacSecret) {
        this.vaultAddr = vaultAddr;
        this.roleId = roleId;
        this.secretId = secretId;
        this.otpHmacKey = new SecretKeySpec(otpHmacSecret, "HmacSHA256");
    }

    // --- Vault authentication (AppRole) ---

    private synchronized String getVaultToken() throws IOException {
        long now = System.currentTimeMillis();
        if (cachedVaultToken != null && now < tokenExpiryEpochMs - 30_000) {
            return cachedVaultToken;
        }

        Map<String, String> payload = new HashMap<>();
        payload.put("role_id", roleId);
        payload.put("secret_id", secretId);

        JsonNode response = httpPost(vaultAddr + "/v1/auth/approle/login", payload, null);
        JsonNode auth = response.get("auth");
        cachedVaultToken = auth.get("client_token").asText();
        long leaseDurationSec = auth.get("lease_duration").asLong();
        tokenExpiryEpochMs = now + (leaseDurationSec * 1000);

        return cachedVaultToken;
    }

    // --- Fetch SMS provider credentials from Vault, with short-lived caching ---

    private synchronized void ensureProviderCredentials() throws IOException {
        long now = System.currentTimeMillis();
        if (cachedAccountSid != null && now < credsExpiryEpochMs) {
            return; // still fresh
        }

        String token = getVaultToken();
        JsonNode response = httpGet(vaultAddr + "/v1/secret/data/sms-provider", token);
        JsonNode data = response.path("data").path("data");

        cachedAccountSid = data.get("account_sid").asText();
        cachedAuthToken = data.get("auth_token").asText();
        cachedFromNumber = data.get("from_number").asText();
        credsExpiryEpochMs = now + CREDS_CACHE_TTL_MS;

        // SAFE: log only that a refresh happened, never the values
        logger.info("SMS provider credentials refreshed from Vault");
    }

    // --- OTP generation, sending, verification ---

    public void generateAndSendOtp(String phoneNumber) throws IOException {
        String otp = String.format("%06d", secureRandom.nextInt(1_000_000));
        String otpHash = hmac(otp);

        otpStore.put(phoneNumber, new OtpRecord(otpHash, Instant.now().plusSeconds(OTP_TTL_SECONDS)));

        sendSms(phoneNumber, "Your verification code is: " + otp);

        // SAFE: never log the OTP itself
        logger.info("OTP generated and sent for phone=" + maskPhone(phoneNumber));
    }

    public boolean verifyOtp(String phoneNumber, String submittedOtp) {
        OtpRecord record = otpStore.get(phoneNumber);

        if (record == null) {
            logger.warning("OTP verification attempted with no active OTP, phone=" + maskPhone(phoneNumber));
            return false;
        }

        if (Instant.now().isAfter(record.expiresAt)) {
            otpStore.remove(phoneNumber);
            logger.info("Expired OTP verification attempt, phone=" + maskPhone(phoneNumber));
            return false;
        }

        if (++record.attempts > MAX_ATTEMPTS) {
            otpStore.remove(phoneNumber);
            logger.warning("OTP attempt limit exceeded, phone=" + maskPhone(phoneNumber));
            return false;
        }

        String submittedHash = hmac(submittedOtp);
        boolean matches = MessageDigest.isEqual(
                submittedHash.getBytes(StandardCharsets.UTF_8),
                record.otpHash.getBytes(StandardCharsets.UTF_8));

        if (matches) {
            otpStore.remove(phoneNumber); // single-use
        }

        logger.info("OTP verification " + (matches ? "succeeded" : "failed")
                + " for phone=" + maskPhone(phoneNumber));
        return matches;
    }

    private void sendSms(String toNumber, String body) throws IOException {
        ensureProviderCredentials();

        String urlStr = "https://api.twilio.com/2010-04-01/Accounts/" + cachedAccountSid + "/Messages.json";
        HttpURLConnection conn = (HttpURLConnection) new URL(urlStr).openConnection();
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);

        String auth = cachedAccountSid + ":" + cachedAuthToken;
        String encodedAuth = Base64.getEncoder().encodeToString(auth.getBytes(StandardCharsets.UTF_8));
        conn.setRequestProperty("Authorization", "Basic " + encodedAuth);
        conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");

        String params = "To=" + toNumber + "&From=" + cachedFromNumber + "&Body=" +
                java.net.URLEncoder.encode(body, "UTF-8");
        try (OutputStream os = conn.getOutputStream()) {
            os.write(params.getBytes(StandardCharsets.UTF_8));
        }

        int status = conn.getResponseCode();
        if (status >= 400) {
            // SAFE: log failure without leaking credentials or OTP content
            logger.warning("SMS send failed with HTTP " + status + " for phone=" + maskPhone(toNumber));
            throw new IOException("SMS provider returned HTTP " + status);
        }
    }

    // --- Helpers ---

    private String hmac(String value) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(otpHmacKey);
            byte[] result = mac.doFinal(value.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(result);
        } catch (Exception e) {
            throw new RuntimeException("OTP hashing failed", e);
        }
    }

    private String maskPhone(String phone) {
        if (phone.length() < 4) return "***";
        return "***" + phone.substring(phone.length() - 4);
    }

    private JsonNode httpPost(String urlStr, Object payload, String token) throws IOException {
        HttpURLConnection conn = (HttpURLConnection) new URL(urlStr).openConnection();
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);
        conn.setRequestProperty("Content-Type", "application/json");
        if (token != null) conn.setRequestProperty("X-Vault-Token", token);
        try (OutputStream os = conn.getOutputStream()) {
            os.write(mapper.writeValueAsBytes(payload));
        }
        return readResponse(conn);
    }

    private JsonNode httpGet(String urlStr, String token) throws IOException {
        HttpURLConnection conn = (HttpURLConnection) new URL(urlStr).openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("X-Vault-Token", token);
        return readResponse(conn);
    }

    private JsonNode readResponse(HttpURLConnection conn) throws IOException {
        int status = conn.getResponseCode();
        InputStream stream = (status >= 200 && status < 300) ? conn.getInputStream() : conn.getErrorStream();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
            if (status >= 200 && status < 300) {
                return mapper.readTree(sb.toString());
            }
            throw new IOException("Vault request failed (HTTP " + status + "): " + sb);
        }
    }

    public static void main(String[] args) throws IOException {
        String vaultAddr = System.getenv("VAULT_ADDR");
        String roleId = System.getenv("VAULT_ROLE_ID");
        String secretId = System.getenv("VAULT_SECRET_ID");
        byte[] otpHmacSecret = System.getenv("OTP_HMAC_SECRET").getBytes(StandardCharsets.UTF_8);

        SecureSmsAuthService service = new SecureSmsAuthService(vaultAddr, roleId, secretId, otpHmacSecret);
        service.generateAndSendOtp("+15559876543");
    }
}